File: plugins/app/src/main/java/com/mapbox/mapboxsdk/plugins/testapp/activity/BuildingActivity.java
package com.mapbox.mapboxsdk.plugins.testapp.activity;

import android.graphics.Color;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.SeekBar;

import com.mapbox.androidsdk.plugins.building.BuildingPlugin;
import com.mapbox.mapboxsdk.constants.Style;
import com.mapbox.mapboxsdk.maps.MapView;
import com.mapbox.mapboxsdk.maps.MapboxMap;
import com.mapbox.mapboxsdk.maps.OnMapReadyCallback;
import com.mapbox.mapboxsdk.plugins.testapp.R;
import com.mapbox.mapboxsdk.style.light.Light;
import com.mapbox.mapboxsdk.style.light.Position;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import timber.log.Timber;

/**
 * Activity showcasing building plugin integration
 */
public class BuildingActivity extends AppCompatActivity implements OnMapReadyCallback {

  @BindView(R.id.mapView)
  MapView mapView;

  @BindView(R.id.fabBuilding)
  View fab;

  @BindView(R.id.seekbarLightRadialCoordinate)
  SeekBar seekbarRadialCoordinate;

  @BindView(R.id.seekbarLightAzimuthalAngle)
  SeekBar seekbarAzimuthalAngle;

  @BindView(R.id.seekbarLightPolarAngle)
  SeekBar seekbarPolarAngle;

  private MapboxMap mapboxMap;
  private BuildingPlugin buildingPlugin;
  private boolean isEnabled;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_building);
    ButterKnife.bind(this);
    mapView.onCreate(savedInstanceState);
    mapView.getMapAsync(this);
  }

  @Override
  public void onMapReady(MapboxMap mapboxMap) {
    this.mapboxMap = mapboxMap;
    buildingPlugin = new BuildingPlugin(mapView, mapboxMap);
    buildingPlugin.setMinZoomLevel(15);
    fab.setVisibility(View.VISIBLE);
    initLightSeekbar();
  }

  private void initLightSeekbar() {
    seekbarRadialCoordinate.setMax(24); // unknown?
    seekbarAzimuthalAngle.setMax(180); // unknown?
    seekbarPolarAngle.setMax(180); // polar angle ranges from 0 to 180 degrees

    PositionChangeListener positionChangeListener = new PositionChangeListener();
    seekbarRadialCoordinate.setOnSeekBarChangeListener(positionChangeListener);
    seekbarAzimuthalAngle.setOnSeekBarChangeListener(positionChangeListener);
    seekbarPolarAngle.setOnSeekBarChangeListener(positionChangeListener);
  }

  @OnClick(R.id.fabBuilding)
  public void onBuildingFabClicked() {
    if (mapboxMap != null) {
      isEnabled = !isEnabled;
      buildingPlugin.setVisibility(isEnabled);
      Timber.e("Building plugin is enabled :%s", isEnabled);
    }
  }

  private class PositionChangeListener implements SeekBar.OnSeekBarChangeListener {
    @Override
    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
      invalidateLightPosition();
    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {
      // Only listening to positionChange for onProgress.
    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {
      // Only listening to positionChange for onProgress.
    }
  }

  private void invalidateLightPosition() {
    Light light = mapboxMap.getLight();
    float radialCoordinate = (float) seekbarRadialCoordinate.getProgress() / 20;
    float azimuthalAngle = seekbarAzimuthalAngle.getProgress();
    float polarAngle = seekbarPolarAngle.getProgress();
    light.setPosition(new Position(radialCoordinate, azimuthalAngle, polarAngle));
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    switch (item.getItemId()) {
      case R.id.menu_building_min_zoom:
        buildingPlugin.setMinZoomLevel(14);
        return true;
      case R.id.menu_building_opacity:
        buildingPlugin.setOpacity(1.0f);
        return true;
      case R.id.menu_building_color:
        buildingPlugin.setColor(Color.RED);
        return true;
      case R.id.menu_building_style:
        mapboxMap.setStyle(Style.DARK);
        return true;
      default:
        return super.onOptionsItemSelected(item);
    }
  }

  @Override
  public boolean onCreateOptionsMenu(Menu menu) {
    getMenuInflater().inflate(R.menu.menu_building, menu);
    return super.onCreateOptionsMenu(menu);
  }

  @Override
  protected void onStart() {
    super.onStart();
    mapView.onStart();
  }

  @Override
  protected void onResume() {
    super.onResume();
    mapView.onResume();
  }

  @Override
  protected void onPause() {
    super.onPause();
    mapView.onPause();
  }

  @Override
  protected void onStop() {
    super.onStop();
    mapView.onStop();
  }

  @Override
  protected void onSaveInstanceState(Bundle outState) {
    super.onSaveInstanceState(outState);
    mapView.onSaveInstanceState(outState);
  }

  @Override
  protected void onDestroy() {
    super.onDestroy();
    mapView.onDestroy();
  }

  @Override
  public void onLowMemory() {
    super.onLowMemory();
    mapView.onLowMemory();
  }
}

File: app/src/main/java/com/mapbox/mapboxsdk/plugins/testapp/activity/building/BuildingActivity.kt
package com.mapbox.mapboxsdk.plugins.testapp.activity.building

import android.graphics.Color
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.SeekBar

import com.mapbox.mapboxsdk.constants.Style
import com.mapbox.mapboxsdk.maps.MapView
import com.mapbox.mapboxsdk.maps.MapboxMap
import com.mapbox.mapboxsdk.maps.OnMapReadyCallback
import com.mapbox.mapboxsdk.plugins.building.BuildingPlugin
import com.mapbox.mapboxsdk.plugins.testapp.R
import com.mapbox.mapboxsdk.style.light.Position

import kotlinx.android.synthetic.main.activity_building.*
import timber.log.Timber

/**
 * Activity showcasing building plugin integration
 */
class BuildingActivity : AppCompatActivity(), OnMapReadyCallback {

    private var mapboxMap: MapboxMap? = null
    private var buildingPlugin: BuildingPlugin? = null
    private var isEnabled: Boolean = false
    lateinit var mapView: MapView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_building)
        this.mapView = findViewById(R.id.mapView)
        mapView.onCreate(savedInstanceState)
        mapView.getMapAsync(this)

        fabBuilding.setOnClickListener{ _ ->
            buildingPlugin?.let {
                isEnabled = !isEnabled
                it.setVisibility(isEnabled)
                Timber.e("Building plugin is enabled :%s", isEnabled)
            }
        }
    }

    override fun onMapReady(mapboxMap: MapboxMap) {
        this.mapboxMap = mapboxMap
        buildingPlugin = BuildingPlugin(mapView, mapboxMap)
        buildingPlugin?.setMinZoomLevel(15f)
        fabBuilding.visibility = View.VISIBLE
        initLightSeekbar()
    }

    // See https://en.wikipedia.org/wiki/Spherical_coordinate_system for more information on these values
    private fun initLightSeekbar() {
        seekbarLightRadialCoordinate.max = 24 // unknown?
        seekbarLightAzimuthalAngle.max = 180 // unknown?
        seekbarLightPolarAngle.max = 180 // polar angle ranges from 0 to 180 degrees

        val positionChangeListener = PositionChangeListener()
        seekbarLightRadialCoordinate.setOnSeekBarChangeListener(positionChangeListener)
        seekbarLightAzimuthalAngle.setOnSeekBarChangeListener(positionChangeListener)
        seekbarLightPolarAngle.setOnSeekBarChangeListener(positionChangeListener)
    }

    private inner class PositionChangeListener : SeekBar.OnSeekBarChangeListener {
        override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
            invalidateLightPosition()
        }

        override fun onStartTrackingTouch(seekBar: SeekBar) {
            // Only listening to positionChange for onProgress.
        }

        override fun onStopTrackingTouch(seekBar: SeekBar) {
            // Only listening to positionChange for onProgress.
        }
    }

    private fun invalidateLightPosition() {
        val light = mapboxMap?.light
        val radialCoordinate = seekbarLightRadialCoordinate.progress.toFloat() / 20
        val azimuthalAngle = seekbarLightAzimuthalAngle.progress.toFloat()
        val polarAngle = seekbarLightPolarAngle.progress.toFloat()
        light?.position = Position(radialCoordinate, azimuthalAngle, polarAngle)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_building_min_zoom -> {
                buildingPlugin?.setMinZoomLevel(14f)
                return true
            }
            R.id.menu_building_opacity -> {
                buildingPlugin?.setOpacity(1.0f)
                return true
            }
            R.id.menu_building_color -> {
                buildingPlugin?.setColor(Color.RED)
                return true
            }
            R.id.menu_building_style -> {
                mapboxMap?.setStyle(Style.DARK)
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_building, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onStart() {
        super.onStart()
        mapView.onStart()
    }

    override fun onResume() {
        super.onResume()
        mapView.onResume()
    }

    override fun onPause() {
        super.onPause()
        mapView.onPause()
    }

    override fun onStop() {
        super.onStop()
        mapView.onStop()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        mapView.onSaveInstanceState(outState)
    }

    override fun onDestroy() {
        super.onDestroy()
        mapView.onDestroy()
    }

    override fun onLowMemory() {
        super.onLowMemory()
        mapView.onLowMemory()
    }

    fun getBuildingPlugin(): BuildingPlugin? {
        return buildingPlugin
    }
}
